### PR TITLE
Ignore fuchsia device-finder no device error

### LINKED ISSUE #51806: Installing Fuchsia SDK causes "device-finder failed: no devices found" errors even when not targeting Fuchsia
`device-finder list` is exiting with a non-zero code when it can't find any devices, which is causing an error message to be printed.

```
$ flutter precache --fuchsia --no-android --no-ios --force
$ flutter devices                                                                 
device-finder failed: 2020/03/02 16:20:04 no devices found


2 connected devices:

Android SDK built for x86 • emulator-5554                        • android-x86 • Android 10 (API 29) (emulator)
iPhone 11 Pro Max         • 90DF02A0-1C0A-432E-9E4F-C5C4A530F323 • ios         • com.apple.CoreSimulator.SimRuntime.iOS-13-4 (simulator)
```
(device-finder failed message is red error text).  Also reproduces with `flutter run` when I am not trying to run against a Fuchsia device.

There should only be an error if a missing device is targeted.

### PR DESCRIPTION
**Pull request by jmagman**
## Description

Do not log "list.go:72: no devices found" when no fuchsia devices are found, just return null like the other device finders.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/51806

## Tests

fuchsia_dev_finder_test.dart

## Checklist
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change
- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

### CODE REVIEW

File: packages/flutter_tools/lib/src/fuchsia/fuchsia_dev_finder.dart [hunk 1]
```diff
@@ -19,24 +22,42 @@ import 'fuchsia_sdk.dart';
...
       if (timeout != null)
         ...<String>['-timeout', '${timeout.inMilliseconds}ms']
     ];
-    final RunResult result = await processUtils.run(command);
+    final RunResult result = await _processUtils.run(command);
     if (result.exitCode != 0) {
-      globals.printError('device-finder failed: ${result.stderr}');
+      // No devices returns error code 1.
+      // https://bugs.fuchsia.dev/p/fuchsia/issues/detail?id=48563
+      if (!result.stderr.contains('no devices found')) {
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/fuchsia/fuchsia_dev_finder.dart]: This is the only behavior change, the rest of this diff is making `FuchsiaDevFinder` more testable.

File: packages/flutter_tools/test/general.shard/fuchsia/fuchsia_dev_finder_test.dart [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        );
+
+        expect(await fuchsiaDevFinder.list(timeout: const Duration(seconds: 2)), <String>['device1']);
+      });
+    });
+  });
+}
+
+class MockFuchsiaArtifacts extends Mock implements FuchsiaArtifacts {}
+class MockProcessManager extends Mock implements ProcessManager {}
```
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/fuchsia/fuchsia_dev_finder_test.dart]: nit: Missing newline at end-of-file

**Review by zanderso — APPROVED** [review]
Oh, man, we need to get the Fuchsia support guarded by a `flutter config` `Feature` rather than gating on the presence of the Fuchsia SDK.

lgtm with nits. thanks for doing this!

File: packages/flutter_tools/test/general.shard/fuchsia/fuchsia_dev_finder_test.dart [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+void main() {
+  MockFuchsiaArtifacts mockFuchsiaArtifacts;
+  BufferLogger logger;
+  MemoryFileSystem memoryFileSystem;
+  File deviceFinder;
+
+  setUp(() {
+    mockFuchsiaArtifacts = MockFuchsiaArtifacts();
+    memoryFileSystem = MemoryFileSystem();
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/fuchsia/fuchsia_dev_finder_test.dart]: use `MemoryFileSystem.test()`

File: packages/flutter_tools/test/general.shard/fuchsia/fuchsia_dev_finder_test.dart [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+      expect(await fuchsiaDevFinder.list(), isNull);
+      expect(logger.errorText, isEmpty);
+    });
+
+    testWithoutContext('error', () async {
+      deviceFinder.createSync();
+
+      final ProcessManager processManager = FakeProcessManager.list(<FakeCommand>[
+        FakeCommand(
+          command: <String>[ deviceFinder.path, 'list', '-full',
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/fuchsia/fuchsia_dev_finder_test.dart]: nit: odd formatting

File: packages/flutter_tools/test/general.shard/fuchsia/fuchsia_dev_finder_test.dart [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+      expect(await fuchsiaDevFinder.list(), isNull);
+      expect(logger.errorText, contains('unexpected error'));
+    });
+
+    testWithoutContext('devices found', () async {
+      deviceFinder.createSync();
+
+      final ProcessManager processManager = FakeProcessManager.list(<FakeCommand>[
+        FakeCommand(
+          command: <String>[ deviceFinder.path, 'list', '-full',
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/fuchsia/fuchsia_dev_finder_test.dart]: same here

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Pull request merged**